### PR TITLE
Persist JobRun.Overrides in case node goes down

### PR DESCRIPTION
**Pull request by dimroc**
https://www.pivotaltracker.com/story/show/158678025

### CODE REVIEW

File: services/job_runner_test.go [hunk 1]
```diff
@@ -213,6 +213,25 @@ func TestJobRunner_ExecuteRun(t *testing.T) {
...
+	job.Tasks = []models.TaskSpec{}
+	run := job.NewRun(initr)
+
+	initialData := models.JSON{Result: gjson.Parse(`{"key":"shouldBeHere"}`)}
+	overrides := models.RunResult{Data: initialData}
+	run, err := services.ExecuteRunAtBlock(run, store, overrides, nil)
+	assert.Error(t, err)
+
+	store.One("ID", run.ID, &run)
+	assert.Equal(t, initialData, run.Overrides.Data)
```
Comment (rupurt) [inline services/job_runner_test.go]: Do you think we should also add the negative assertion that it doesn't save overrides on success?

File: services/job_runner_test.go [hunk 2]
```diff
@@ -213,6 +213,25 @@ func TestJobRunner_ExecuteRun(t *testing.T) {
...
+	job, initr := cltest.NewJobWithLogInitiator()
+	job.Tasks = []models.TaskSpec{}
+	run := job.NewRun(initr)
+
+	initialData := models.JSON{Result: gjson.Parse(`{"key":"shouldBeHere"}`)}
+	overrides := models.RunResult{Data: initialData}
+	run, err := services.ExecuteRunAtBlock(run, store, overrides, nil)
+	assert.Error(t, err)
+
+	store.One("ID", run.ID, &run)
```
Comment (se3000) [inline services/job_runner_test.go]: Also not crucial, but it'd be nice to assert no error happens here.

File: services/job_runner.go [hunk 3]
```diff
@@ -214,37 +204,53 @@ func ExecuteRunAtBlock(
...
-	jr = jr.ApplyResult(latestRun.Result)
+	jr = jr.ApplyResult(prevResult)
 	logger.Infow("Finished current job run execution", jr.ForLogger()...)
 	return jr, wrapError(jr, store.Save(&jr))
 }
 
+func prepareJobRun(jr models.JobRun, store *store.Store, overrides models.RunResult, bn *models.IndexableBlockNumber) (models.JobRun, error) {
+	if jr.Status.CanStart() {
+		jr.Status = models.RunStatusInProgress
+	}
```
Comment (se3000) [inline services/job_runner.go]: I get that this is unchanged from the old code, but while we're touching it: shouldn't we exit immediately if we can't start?

File: services/job_runner.go [hunk 4]
```diff
@@ -214,37 +204,53 @@ func ExecuteRunAtBlock(
...
+	var err error
+	jr.Overrides, err = jr.Overrides.Merge(overrides)
+	if err != nil {
+		return jr, err
```
Comment (se3000) [inline services/job_runner.go]: Seems like we may want to set the job run status to errored here if we can't merge parameters.

File: services/job_runner_test.go (hunk 1 above)
Comment (dimroc) [reply services/job_runner_test.go]: I am under the impression we want to save the overrides on success or failure. I could assert for it's presence in the successful scenario as well.

File: services/job_runner.go (hunk 4 above)
Comment (dimroc) [reply services/job_runner.go]: I agree. But I also think we should set the job status to errored if anything in this `prepareJobRun` errors and floated the idea of having a `defer` handle all errors inside `ExecuteJobRunAtBlock`. If we're fine with this particular scenario for now I'll add it. But I thought it awkward to cherry pick this error as bubbling up to the job run and none of the others.

File: services/job_runner.go [hunk 5]
```diff
@@ -22,38 +23,48 @@ type JobRunner interface {
...
 func (rm *jobRunner) Stop() {
-	rm.closingMutex.Lock()
-	defer rm.closingMutex.Unlock()
+	rm.bootMutex.Lock()
+	defer rm.bootMutex.Unlock()
 
-	if !rm.closed {
-		rm.closed = true
-		close(rm.closer)
+	if !rm.started {
```
Comment (j16r) [inline services/job_runner.go]: Seems like `bootMutex` and `started` could be an atomic rather than a mutex + bool?

Comment (dimroc) [reply services/job_runner.go]: The bootMutex is protecting more than just the bool, conceptually I see it as protecting the whole `start()` method. The creation and destruction of `rm.done` is another example of something we don't want accessed in parallel. But perhaps I'm being overprotective.

**Pull request merged**